File: src/worker/core/season/genPlayoffSeeds.cpp
#include <vector>
#include <stdexcept>
#include <cmath>

using Seed = std::pair<int, std::optional<int>>; // Return the seeds (0 indexed) for the matchups, in order (std::nullopt is a bye)

std::vector<Seed> genPlayoffSeeds(int numPlayoffTeams, int numPlayoffByes) {
    double numRounds = std::log2(numPlayoffTeams + numPlayoffByes);

    if (numRounds != static_cast<int>(numRounds)) {
        throw std::invalid_argument("Invalid genSeeds input: " + std::to_string(numPlayoffTeams) + " teams and " + std::to_string(numPlayoffByes) + " byes");
    }

    // Handle byes - replace lowest seeds with undefined
    std::vector<int> byeSeeds;

    for (int i = 0; i < numPlayoffByes; i++) {
        byeSeeds.push_back(numPlayoffTeams + i);
    }

    auto addMatchup = [&](std::vector<Seed>& currentRound, std::optional<int> team1, int maxTeamInRound) {
        if (!team1.has_value()) {
            throw std::invalid_argument("Invalid type");
        }

        int otherTeam = maxTeamInRound - team1.value();
        currentRound.emplace_back(team1.value(), std::find(byeSeeds.begin(), byeSeeds.end(), otherTeam) != byeSeeds.end() ? std::nullopt : otherTeam);
    };

    // Grow from the final matchup
    std::vector<Seed> lastRound = { {0, 1} };

    for (int i = 0; i < numRounds - 1; i++) {
        const int numTeamsInRound = static_cast<int>(lastRound.size()) * 4;
        const int maxTeamInRound = numTeamsInRound - 1;
        std::vector<Seed> currentRound;

        for (const auto& matchup : lastRound) {
            // swapOrder stuff is just cosmetic, matchups would be the same without it, just displayed slightly differently
            bool swapOrder = (currentRound.size() / 2) % 2 == 1 && matchup.second.has_value();
            addMatchup(currentRound, swapOrder ? matchup.second : matchup.first, maxTeamInRound);
            addMatchup(currentRound, swapOrder ? matchup.first : matchup.second, maxTeamInRound);
        }

        lastRound = currentRound;
    }

    return lastRound;
}

File: src/worker/core/season/genPlayoffSeeds.rs
type Seed = (i32, Option<i32>); // Return the seeds (0 indexed) for the matchups, in order (None is a bye)

fn gen_playoff_seeds(num_playoff_teams: i32, num_playoff_byes: i32) -> Vec<Seed> {
    let num_rounds = (num_playoff_teams + num_playoff_byes) as f64).log2();

    if num_rounds.fract() != 0.0 {
        panic!(
            "Invalid genSeeds input: {} teams and {} byes",
            num_playoff_teams, num_playoff_byes
        );
    }

    // Handle byes - replace lowest seeds with None
    let mut bye_seeds: Vec<i32> = Vec::new();

    for i in 0..num_playoff_byes {
        bye_seeds.push(num_playoff_teams + i);
    }

    fn add_matchup(current_round: &mut Vec<Seed>, team1: Option<i32>, max_team_in_round: i32) {
        let team1_value = team1.expect("Invalid type");

        let other_team = max_team_in_round - team1_value;
        current_round.push((
            team1_value,
            if bye_seeds.contains(&other_team) { None } else { Some(other_team) },
        ));
    }

    // Grow from the final matchup
    let mut last_round: Vec<Seed> = vec![(0, Some(1))];

    for _ in 0..(num_rounds as i32 - 1) {
        // Add two matchups to currentRound, for the two teams in lastRound. The sum of the seeds in a matchup is constant for an entire round!
        let num_teams_in_round = last_round.len() as i32 * 4;
        let max_team_in_round = num_teams_in_round - 1;
        let mut current_round: Vec<Seed> = Vec::new();

        for matchup in &last_round {
            // swap_order stuff is just cosmetic, matchups would be the same without it, just displayed slightly differently
            let swap_order = (current_round.len() / 2) % 2 == 1 && matchup.1.is_some();
            add_matchup(&mut current_round, if swap_order { matchup.1 } else { matchup.0 }, max_team_in_round);
            add_matchup(&mut current_round, if swap_order { matchup.0 } else { matchup.1 }, max_team_in_round);
        }

        last_round = current_round;
    }

    last_round
}
